File: IntelliJ IDEA/Java/src/com/eruru/mvvm/swing/MVVMItemsControl.java
package com.eruru.mvvm.swing;

import com.eruru.mvvm.*;

import javax.swing.*;
import java.awt.*;

public class MVVMItemsControl extends MVVMControl {

	private final Container control;

	private MVVMBinding itemsSource;
	private MVVMBinding itemTemplate;
	private MVVMNotifyCollectionChanged notifyCollectionChanged;

	public MVVMItemsControl (Container control) {
		super (control);
		this.control = control;
	}

	public Container getControl () {
		return control;
	}

	public MVVMBinding getItemsSource () {
		return getBinding (itemsSource, this::setItemsSource);
	}

	public void setItemsSource (MVVMBinding binding) {
		itemsSource = setBinding (binding, targetValue -> {
			cancelCollectionChanged ();
			reset ();
			if (targetValue instanceof Iterable) {
				for (Object item : (Iterable<?>) targetValue) {
					add (item);
				}
			}
			if (targetValue instanceof MVVMNotifyCollectionChanged) {
				notifyCollectionChanged = (MVVMNotifyCollectionChanged) targetValue;
				notifyCollectionChanged.getCollectionChanged ().add (this, this::collectionChanged);
			}
		}, this::cancelCollectionChanged);
	}

	public MVVMBinding getItemTemplate () {
		return getBinding (itemTemplate, this::setItemTemplate);
	}

	public void setItemTemplate (MVVMBinding binding) {
		itemTemplate = setBinding (binding);
	}

	protected Object convert (Object value) {
		MVVMDataTemplate dataTemplate = null;
		if (itemTemplate != null) {
			dataTemplate = itemTemplate.getTargetValue (MVVMDataTemplate.class);
		}
		MVVMControlBase control;
		if (dataTemplate == null) {
			control = new MVVMLabel (new JLabel ()) {{
				setText (new MVVMBinding ());
			}};
		} else {
			control = dataTemplate.build ();
		}
		return control;
	}

	private MVVMControl register (Object value, int index) {
		MVVMControl control = (MVVMControl) convert (value);
		super.add (control, index);
		control.setDataContext (new MVVMBinding (value));
		control.build (getRoot ());
		this.control.add (control.getControl (), index);
		this.control.revalidate ();
		this.control.repaint ();
		return control;
	}

	protected void add (Object value, int index) {
		register (value, index);
	}

	protected void add (Object value) {
		add (value, super.getControls ().size ());
	}

	protected void removeAt (int index) {
		super.remove (index);
		control.remove (index);
		control.revalidate ();
		control.repaint ();
	}

	protected void replace (int index, Object value) {
		removeAt (index);
		register (value, index);
	}

	public void move (int oldIndex, int newIndex) {
		super.move (oldIndex, newIndex);
		control.add (control.getComponent (oldIndex), newIndex);
		control.revalidate ();
		control.repaint ();
	}

	protected void reset () {
		super.clear ();
		control.removeAll ();
		control.revalidate ();
		control.repaint ();
	}

	void cancelCollectionChanged () {
		if (notifyCollectionChanged != null) {
			notifyCollectionChanged.getCollectionChanged ().remove (this);
			notifyCollectionChanged = null;
		}
	}

	void collectionChanged (Object sender, MVVMNotifyCollectionChangedAction action, int oldIndex, Object[] oldItems, int newIndex, Object[] newItems) {
		switch (action) {
			case Add:
				this.add (newItems[0], newIndex);
				break;
			case Remove:
				this.removeAt (oldIndex);
				break;
			case Replace:
				this.replace (oldIndex, newItems[0]);
				break;
			case Move:
				this.move (oldIndex, newIndex);
				break;
			case Reset:
				this.reset ();
				break;
			default:
				throw new UnsupportedOperationException (action.toString ());
		}
	}

}

File: Visual Studio/C#/Eruru.MVVM/Shared/Controls/MVVMItemsControl.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Eruru.MVVM {

	public partial class MVVMItemsControl : MVVMControl {

		public MVVMBinding ItemsSource {

			get {
				return GetBinding (ref _ItemsSource, value => ItemsSource = value);
			}

			set {
				SetBinding (ref _ItemsSource, value, setTargetValue: targetValue => {
					CancelCollectionChanged ();
					Reset ();
					if (targetValue is IEnumerable) {
						AddRange ((IEnumerable)targetValue);
					}
					NotifyCollectionChanged = targetValue as IMVVMNotifyCollectionChanged;
					if (NotifyCollectionChanged != null) {
						NotifyCollectionChanged.CollectionChanged += MVVMItemsControlBase_CollectionChanged;
					}
				}, unbind: () => {
					Reset ();
					CancelCollectionChanged ();
				});
			}

		}
		public MVVMBinding ItemTemplate {

			get {
				return GetBinding (ref _ItemTemplate, value => ItemTemplate = value);
			}

			set {
				SetBinding (ref _ItemTemplate, value);
			}

		}

		protected MVVMFunc<MVVMControl> OnUseDefaultDataTemplate;
		protected MVVMAction<int, MVVMControl> OnInsert;
		protected Action<int> OnRemoveAt;
		protected MVVMAction<int, int> OnMove;
		protected MVVMAction OnReset;

		readonly List<MVVMControl> Items = new List<MVVMControl> ();

		MVVMBinding _ItemsSource;
		MVVMBinding _ItemTemplate;
		IMVVMNotifyCollectionChanged NotifyCollectionChanged;

		protected virtual object Convert (object value) {
			MVVMDataTemplate dataTemplate = _ItemTemplate == null ? null : (_ItemTemplate.GetTargetValue () as MVVMDataTemplate);
			MVVMControl control;
			if (dataTemplate == null) {
				control = OnUseDefaultDataTemplate ();
			} else {
				control = dataTemplate.Build ();
			}
			return control;
		}

		protected virtual void Insert (int index, object value) {
			MVVMControl control = (MVVMControl)Convert (value);
			control.Parent = this;
			control.DataContext = new MVVMBinding (value);
			control.Build (Root);
			Items.Insert (index, control);
			OnInsert (index, control);
		}

		protected virtual void InsertRange (int index, IEnumerable collection) {
			int i = index;
			foreach (object value in collection) {
				Insert (i, value);
				i++;
			}
		}

		protected virtual void Add (object value) {
			Insert (Items.Count, value);
		}

		protected virtual void AddRange (IEnumerable collection) {
			foreach (object value in collection) {
				Add (value);
			}
		}

		protected new virtual void RemoveAt (int index) {
			Items[index].Unbind ();
			Items.RemoveAt (index);
			OnRemoveAt (index);
		}

		protected virtual void Replace (int index, object value) {
			RemoveAt (index);
			Insert (index, value);
		}

		protected new virtual void Move (int oldIndex, int newIndex) {
			MVVMControl control = Items[oldIndex];
			Items.RemoveAt (oldIndex);
			Items.Insert (newIndex, control);
			OnMove (oldIndex, newIndex);
		}

		protected virtual void Reset () {
			foreach (MVVMControl control in Items) {
				control.Unbind ();
			}
			Items.Clear ();
			OnReset ();
		}

		void MVVMItemsControlBase_CollectionChanged (object sender, MVVMNotifyCollectionChangedEventArgs e) {
			//Console.WriteLine ("CollectionChanged {0} from {1} to {2}", e.Action, sender, ItemsSource);
			switch (e.Action) {
				case MVVMNotifyCollectionChangedAction.Add:
					Insert (e.NewStartingIndex, e.NewItems[0]);
					break;
				case MVVMNotifyCollectionChangedAction.Remove:
					RemoveAt (e.OldStartingIndex);
					break;
				case MVVMNotifyCollectionChangedAction.Replace:
					Replace (e.OldStartingIndex, e.NewItems[0]);
					break;
				case MVVMNotifyCollectionChangedAction.Move:
					Move (e.OldStartingIndex, e.NewStartingIndex);
					break;
				case MVVMNotifyCollectionChangedAction.Reset:
					Reset ();
					break;
				default:
					throw new NotImplementedException (e.Action.ToString ());
			}
		}

		void CancelCollectionChanged () {
			if (NotifyCollectionChanged != null) {
				NotifyCollectionChanged.CollectionChanged -= MVVMItemsControlBase_CollectionChanged;
				NotifyCollectionChanged = null;
			}
		}

	}

}